hastory-server/src/Data/Hastory/Server/TestUtils.hs

```diff
@@ -56,7 +77,44 @@ withTestServer func = do
           void $ runSqlPool (runMigrationSilent migrateAll) siPool
           let mkApp = pure $ app opts settings
               opts = Options 10 Nothing
-              settings = ServerSettings siToken siPool
+              settings = ServerSettings siPool jwtSettings defaultCookieSettings
           testWithApplication mkApp $ \p ->
             let siClientEnv = mkClientEnv manager (BaseUrl Http "127.0.0.1" p "")
              in func (ServerInfo {..})
+
+type RegistrationData = (UserId, Token)
+
+withNewUser :: ClientEnv -> (RegistrationData -> Expectation) -> Expectation
+withNewUser clientEnv func = do
+  userForm <- randomUserForm
+  Right userId <- runClientM (createUserClient userForm) clientEnv
+  Right resp <- runClientM (createSessionClient userForm) clientEnv
+  case extractJWTCookie resp of
+    Nothing -> expectationFailure "JWT Cookie not found"
+    Just cookie -> func (userId, Token cookie)
+
+extractJWTCookie :: Headers AuthCookies NoContent -> Maybe B.ByteString
+extractJWTCookie headerList = do
+  (_, cookie) <- find jwtCookie (getHeaders headerList)
```

Comment:
This is awefully complex and probably wrong.
Have you seen this: https://github.com/NorfairKing/intray/blob/d5ab313b91ccea8ece15982bce509ee49d0aef0b/intray-cli/src/Intray/Cli/Commands/Login.hs

Reply:
@NorfairKing I liked this exercise a lot as well. I don't know why, but I don't reach for pattern matching by default. I think it's because I have this idea in my head that pattern matching destroys polymorphism (since the data constructors commit you to a specific type), but so what, right?